### PR TITLE
Backend renaming (1.0 version bumps plus the lexicon)

### LINKED ISSUE #233: Strict mode is wrong
I learned this morning that I've fundamentally misunderstood the handling and usage of `state.data`, and as a result, strict mode is wrong.

I thought:

The runtime will manage various bits of state (`config`, `references`), but `data` is for your application data. `state.data` is the thing you want forwarded on to the next job, it's the thing you want in your final result.

I also knew that adaptors like `http` will scribble on `data` (ie, the response). Which I never liked because it erodes the idea that `data` is a safe space for the user.

But all this is incorrect. In fact the opposite is true:

The runtime will manage the `configuration`, `references` and `data` properties of `state`. The rest of the object is left alone for the user to do what they want with.

Adaptors don't modify `state.data` , they _replace_ it (via `composeNextState`)

The idea of strict mode was to establish new, possibly breaking, conventions for the v2 runtime. This includes encouraging users to lean more heavily on `state.data` and take ownership of it, but it turns out that doesn't make any sense.

## Options

* Remove strict mode entirely (this may be a wise short-term measure)
* Keep strict mode but change its meaning. Maybe strict mode returns state apart from `configuration`, `references`, `data` - because that's all  transient stuff needed for the job.

I kinda wish we had a property of state which represents the result, the final or persistent state, the thing that SHOULD get passed to the next job. A part of state which we promise we won't touch or interfere with, since we're a bit loosey goosey about a small number of keys.

Some possible rules for strict mode, if we want to keep it:

* At the end of a job, `configuration`, `data` and ` references` are removed (are there any other managed props?)
* * Removing data is kind of a big deal, but at least is forces you to process response data in that job and prepare it for passing down the line.
* * It would also mean adaptors don't have to tidy up after themselves
* All properties beginning with `_` are removed (including on nested structures)
* Introduce a new property on state, maybe `result` or `final` or `persist`, which is the only thing that gets returned or forwarded.

### LINKED ISSUE #483: Worker: use job names from the attempt
When the worker generates an Execution Plan from an Attempt, it uses the job id for each job.

But a job may also have a human readable name (like "get from salesforce")

We should consider:

a) using the job name instead of the id (problematic if two job names happen to be equal)
b) including a name and an id on a job.

At the moment we use a job id in various places:
* If there's an error, we write it to state as `errors[jobId]`. Human readable job ids would help here
* If there are multiple leaf results, we write them all ot the final state as `state[jobId]: { data: {} ]`
* In logging, when a job starts execution

I think we should use the job NAME, not id, but ensure they are unique and add a number to the end if they are not. Or we could throw, I suppose. The changes of identical job names, in practice, are slim. If no job id or name is provided we can generate a name based on the index, as we do now.

Actually I might be even more radical and say that a job as a _name_, not an id. Because name implies human-readable and encouraging human readable job names makes everything better.

### PR DESCRIPTION
**Pull request by josephjclark**
This PR bumps the runtime up to 1.0 and introduces the Lexicon, a central repo for typings and namings.

I have started to enforce stronger type checking on the tests as part of this PR, which has helped to catch any issues in the  change of formats but also resulted in a large diff!

Closes #483 #556 #409 #233

Runtime changes:
- use new-style Execution Plan format
- rename jobs to steps (mostly, where it makes sense)
- remove strict state (closes #233)

CLI changes
- remove strict state (closes #233)
- Support but deprecate old workflow data structures
- Support execution plans
- Better use of step, job and expression in code and output #483 

Engine Changes
- Just an adjustment to match new typings

Worker Changes
- Just an adjustment to match new typings

### CODE REVIEW

**Comment by josephjclark** [comment]
Hi @mtuchi when you have a spare half an hour would you mind taking a look at the CLI on this branch?

There's some stuff about terminology in the CLI readme - probably a good idea to look at that first.

You should look closely at the log output and make sure that it aligns with your documentation. Also if workflows and steps have names, they should be used in the log output now (instead of UUIDS), although that likely affects the worker more than the CLI

File: packages/lexicon/README.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+The OpenFunction stack is built on the concepts of Workflows, Runs, Jobs and Expressions (and more). Some of these terms can be used interchangable, or used differently in certain contexts.
+
+Here are the key concepts
+
+- An **Expression** is a string of Javascript (or Javascript-like code) written to be run in the CLI or Lightning.
+- A **Job** is an expression plus some metadata required to run it - typically an adaptor and credentials.
+  The terms Job and Expression are often used interchangeably.
+- A **Workflow** is a series of steps to be executed in sequence. Steps are usually Jobs (and so job and step are often used
+  interchangeably), but can be Triggers.
```
Comment (taylordowns2000) [inline packages/lexicon/README.md:15]: i think we don't technical refer to triggers as steps. right now, all available "steps" are (javascript-based) jobs. in the future, we might introduce a "notification" step which behaves differently. or even a "loop over array" step which does something else entirely. but for now, all steps are jobs.

File: packages/lexicon/core.d.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+export type ConditionalStepEdge = {
+  condition?: string; // Javascript expression (function body, not function)
+  label?: string;
+  disabled?: boolean;
+};
+
+/**
+ * A no-op type of Step
+ */
+export interface Trigger extends Step {}
```
Comment (taylordowns2000) [inline packages/lexicon/core.d.ts:115]: might be removable, unless this is how it works in worker. (happy to consider triggers as steps if that's helpful)

File: packages/lexicon/lightning.d.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+/**
+ * Type definitions for Lightning and Worker interfaces
+ *
+ * This is the lightning-worker contract
+ *
+ * It is helpful to have these in the lexicon to avoid a circular dependency between lightning and the worker
+ * It's also kinda nice that the contract isn't in the worker itself, it's on neutral ground
+ */
+// An run object returned by Lightning
+export type Run = {
```
Comment (taylordowns2000) [inline packages/lexicon/lightning.d.ts:14]: im a little surprised that this isn't called `Plan`, but i'm reading on to see what comes next!

File: packages/lexicon/lightning.d.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+export type STEP_COMPLETE = 'step:complete';
+
+export type ExitReason = {
+  reason: ExitReasonStrings;
+  error_message: string | null;
+  error_type: string | null;
+};
+
+export type GetPlanPayload = void; // no payload
+export type GetPlanReply = Run;
```
Comment (taylordowns2000) [inline packages/lexicon/lightning.d.ts:112]: right, i remember being a little confused by this. when we "getPlan" i'd expect a plan to be returned... but a "Run" is returned. It's not technically a `Run` in lightning's world. To lightning, it's a `RunPlusSomeOtherThings` - is that not exactly what we want to call a `Plan`?

**Review by taylordowns2000 — COMMENTED** [review]
i've reviewed the lexicon section and added comments inline. nothing is a deal-breaker for me, so happy for this to be the 1.0 and adjust if and when necessary. thanks for this work!

File: packages/lexicon/core.d.ts (hunk 2 above)
Comment (josephjclark) [reply packages/lexicon/core.d.ts:115]: It's useful in the worker for sure - we already recognise "execuable jobs" and "non-executable jobs" (which are triggers). Formalising it like this feels like a nice step to me, and it doesn't block any future step-types from being added

File: packages/lexicon/lightning.d.ts (hunk 4 above)
Comment (josephjclark) [reply packages/lexicon/lightning.d.ts:112]: Well - yes, you're right! I'll rename it to `LightningPlan` though to help distinguish Lighting's `RunPlusSomeOtherThings` structure from the runtime's `ExecutionPlan` structure.

But in the future Lightning will be sending the `ExecutionPlan` directly (I hope!)

**Review by stuartc — APPROVED** [review]
Looks good to me, obviously I didn't read everything - but yeah kudos on the lexicon stuff. 

Please merge it when you'd like!

**Comment by josephjclark** [comment]
Thanks @stuartc !

**Pull request merged**